### ISSUE TITLE
How memory works in PPO with `sequence_length`?

### ISSUE THREAD

**Issue by lcswillems** [Unity-Technologies/ml-agents#853]
Thank you for all the work done in this project! I am trying to understand how the memory works in PPO with the `sequence_length` argument. I tried to read https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/ppo/models.py, https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/models.py and https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/ppo/trainer.py but couldn't still understand.

Let's say the replay memory contains 40 tuples `(o_i, m_i)` (where `o_i` means observation and `m_i` means memory state) and `sequence_length` is 4. Would you separate the memory into chunks of 4, like `[(o_1, m_1), ..., (o_4, ..., m_4)], ..., [(o_37, m_37), ..., (o_40, m_40)]`, then compute the loss for each chunk and average it?

Where should I find the answer to this question in the code?

Thank you in advance for your help!

**Comment by vincentpierre** [comment]
I would start with `create_recurrent_encoder`  line 214 of `python/unitytrainers/models.py`. But you are right, the 40 tuples would be split into 10 sequences of length. The LSTM cell takes as input a matrix with an extra dimension corresponding to the sequence. When you do gradient descent with minibatch, you always average the loss for each experience / sequence.

**Comment by lcswillems** [comment]
Thank you for your response! I think I implemented something similar then, but the memory doesn't seem to work well in my code: it works better when sequence length is 1 rather than 4 or 8 for example. Is it also the case in ML-agents? If not, do you have any idea?

Anyway, thanks a lot for your answer!

**Comment by vincentpierre** [comment]
With a Sequence Length of 1, the agent should not be able to use memories properly. You need to train with an environment that can only be solved when the agent needs to remember something. (The Hallway environment for instance). Maybe try to copy the code of create_recurrent_encoder or even reimplement the method entirely. Using RNN takes longer to train, it could be the case that you did not train the network for long enough. I would look into some NLP tutorials if you are really interested in Recurrent Neural Networks.
I hope this helps.

**Comment by lcswillems** [comment]
Thank you for this detailed answer. I looked at the create_recurrent_encoder method but I can't see any difference with my code... I will look again.

**Comment by eshvk** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity, but if you need additional assistance, feel free to reopen the issue.
